#+title: BDMM-Prime
#+subtitle: Linear birth-death-sampling phylodynamics
#+author: Tim Vaughan

#+options: title:nil

#+ATTR_HTML: :alt BDMM-Prime: Linear birth-death-sampling phylodynamics
[[./logos/bdmm-prime-logo.png]]

#+SETUPFILE: theme.org
#+EXPORT_FILE_NAME: index.html

Source code: https://github.com/tgvaughan/BDMM-Prime

* Introduction
:PROPERTIES:
:CUSTOM_ID: id-1-Introduction
:END:

BDMM-Prime is a BEAST 2 package for performing phylodynamic inference
under a variety of linear birth-death-sampling models with/without
types.

** Comparison with BDMM
:PROPERTIES:
:CUSTOM_ID: id-2-Introduction-Comparison-with-BDMM
:END:

The BDMM-Prime project is a hard fork of the original [[https://github.com/denisekuehnert/bdmm][BDMM project]].
The intention is to extend the functionality of the original package,
while improving its flexibility and ease of use.  It incorporates the
following enhancements:
- an improved BEAUti interface that allows a much more diverse range of analyses to be configured,
- automatic fall-back to analytical solutions for unstructured (single type) analyses,
- use of stochastic mapping for sampling ancestral states,
- a particle filtering algorithm allowing joint sampling of population trajectories,
- a heavily refactored code base intended to make the package easier to use,
  extend and maintain.

As a result of the many changes that were required in making this transition,
BDMM-Prime is completely incompatible with BDMM itself.  Thus the original
package will be maintained separately to ensure that BEAST 2 XMLs and packages
that depend on it remain usable.

* Getting Started
:PROPERTIES:
:CUSTOM_ID: id-1-Getting-Started
:END:


** Installation
:PROPERTIES:
:CUSTOM_ID: id-2-Getting-Started-Installation
:END:

BDMM-Prime requires a working installation of [[https://www.beast2.org][BEAST 2.7]] which can be obtained from
https://www.beast2.org/. The package itself can then be installed via the
built-in package manager in the following way:

1. Open BEAUti.
2. From the =File= menu select =Manage Packages=.
3. Click the =Package repositories= button at the bottom of the dialog box.
4. Click =Add URL= and enter the following repository URL:\\
    https://tgvaughan.github.io/BDMM-Prime/package.xml.\\
    The list of repositories should now look like this:
    [[./figures/package_repo.png]]
5. Close the repository manager.
6. The latest version of BDMM-Prime should now appear in the list of packages:
   [[./figures/package_manager.png]]
   Select it, then click the =Install/Upgrade= button.
7. Close the package manager.

BDMM-Prime should now be available to use on your system. (You will need to
restart BEAUti before you can set up any analyses.)

Aside: Eventually BDMM-Prime will be avaialable from the default BEAST 2
repository.  However, while it is still in development, it must be
installed from the above repository.

** Setting up your first analysis
:PROPERTIES:
:CUSTOM_ID: id-2-Getting-Started-Setting-up-your-first-analysis
:END:

Here we will step through the process of setting up, running and
interpreting a simple BDMM-Prime analysis.  For the purpose of
demonstration, we will focus on applying a multi-type model to simple
two-type epidemiological data set, but the same general approach can
be taken to set up single-type analyses of other data sets too.

BDMM-Prime analyses can be set up in BEAUti in much the same way that
analyses under using other BEAST 2 models such as BDSKY or the
Bayesian Skyline Plot can be set up. The only part unique to
BDMM-Prime is "tree prior" configuration.  Nonetheless, below we
describe the /complete/ process of setting up a small BDMM-Prime
analysis.

/The following should be read like a short tutorial. Steps
to perform on your computer are highlighted like this:/
#+begin_quote
Start BEAUti.
#+end_quote

*Note:* the following assumes some limited familiarity with BEAST 2.
If you've never used this software before, please consider working
through the [[https://taming-the-beast.org/tutorials/Introduction-to-BEAST2/][Introduction to BEAST 2 tutorial]].

*** Loading sequence data into BEAUti
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Loading-sequence-data-into-BEAUti
:END:

For this tutorial, we'll be using the example influenza data which is
installed along with BDMM-Prime. This data, assembled from
publicly-available data downloaded from NCBI GenBank, is the same set
used in [[https://taming-the-beast.org/tutorials/Structured-coalescent][this MultiTypeTree tutorial]]. It consists of an aligned set of
60 H3N2 HA sequences sampled from New Zealand and Hong Kong. To make
finding this data easy, we first select the package working directory.
#+begin_quote
Select "Set working dir" from the File menu, then choose "BDMM-Prime" from the sub-menu.
#+end_quote

To load the data, simply choose "Import alignment" from the File menu
and select one or more FASTA files.

#+begin_quote
Select "Import alignment" from the File menu to display a file
selection dialog box.  Using this, open the ~examples~ subdirectory
and select the file ~h3n2_2deme.fna~.
#+end_quote

The BEAUti window should now look something like the following:
[[./figures/tutorial_seqloaded.png]]

*** Setting up tip dates
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Setting-up-tip-dates
:END:

The sequences in this data set are sampled at different times, so use
the following instructions to import this information into the analysis.
(This step is unnecessary if sequences were collected at the same time,
or near enough given the anticipated scale of the tree.)

#+begin_quote
Select the Tip Dates panel and check the "Use tip dates" option. Then
press the "Auto-configure" button to set the times based on values
extracted from the sequence headers. Finally, use the radio button
and dropdown menu to interpret everything after the last "~_~"
character as a numerical tip time.
#+end_quote

[[./figures/tutorial_tipdates.png]]

*** Setting the site model
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Setting-the-site-model
:END:

For this example, we will use the basic [[https://doi.org/10.1007%2FBF02101694][HKY]] subsitution model, with
equilibrium nucleotide frequencies fixed to the empirical frequencies
of characters in our alignment.  

#+begin_quote
Select the "Site model" panel. Select the HKY model with "empirical" nucleotide frequencies.
#+end_quote

[[./figures/tutorial_sitemodel.png]]

This choice to fix the equilibrium frequencies is made here only
to reduce the computational complexity of the tutorial analysis.
For production analyses, these should probably be estimated, and
Gamma-distributed site-to-site rate heterogeneity should also be used.

*** Setting the clock model
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Setting-the-clock-model
:END:

Since we have serially-sampled data, our analysis requires some kind
of molecular clock.  For the sake of simplicity, we use a strict
clock, setting the initial value of the clock rate to $5\times 10^{-3}$
substitutions per site per year, which is close to the true
value for influenza.

#+begin_quote
Select the "Cock model" panel. Set the "mean clock rate" value to 0.005.
#+end_quote

[[./figures/tutorial_clock.png]]

*** Selecting the tree prior
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Selecting-the-tree-prior
:END:

The tree prior (also known as the phylodynamic likelihood) is the
component of the analysis which BDMM-Prime provides. Here we configure
the particular birth-death model we will use to relate the tree to
the various population-level parameters we'd like to learn about.
#+begin_quote
Select the "Priors" panel.  To choose the BDMM-Prime tree prior, find
the drop-down menu next to ~Tree.t:h3n2_2deme~ and select "BDMMPrime".
#+end_quote

[[./figures/tutorial_treeprior.png]]

*** Setting up the sample types
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Setting-up-the-sample-types
:END:

The next thing we want to do is to specify the locations associated with each of the
samples included in our analysis.
#+begin_quote
Scroll down to end of the tree prior section and find the table associating tree leaves
with locations. Notice that the sample names have the form =ID_Location_Date=. To
extract the locations from the sample names, click the ~Auto-configure~ button,
select ~split on character~, ensure "~_~" is specified as the delimiter, and select
"2" from the ~take group(s)~ drop-down menu.  Once this is complete, press "Ok".
#+end_quote

[[./figures/tutorial_tiptypes.png]]

The table should now be populated with types extracted from the sample names.

[[./figures/tutorial_tiptypes2.png]]

*** Configuring the multi-type model
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Configuring-the-multi-type-model
:END:

Now we get to the main part of the analysis setup, where we decide exactly how
to model the generation of our data. In our case, we want to model a small part
of the global transmission dynamics of H3N2.

The first thing to do is to is to select our desired parameterization
of the birth-death model. In our case, as we're analysing pathogen
sequences in an attempt to reconstruct transmission dynamics, we'll
use the epidemiological parameterization.
#+begin_quote
Expand the tree prior section by clicking the arrow next to ~Tree.t:h3n2~ on the
left-hand side of of the screen.  Then select "Epi Parameterization" from the
drop-down menu at the top of the expanded section.
#+end_quote

[[./figures/tutorial_parameterization.png]]

With this done, the next task is to define the dimensionality and
initial values of the multi-type birth-death parameters. In BDMM-Prime,
each of these parameters is considered a "skyline parameter" which can
change in a piecewise-constant fashion through time.

The first skyline parameter to consider is the Re parameter, representing
the effective reproductive number.  By default parameter values are
assumed to be the same across all types,

#+begin_quote
For the Re skyline parameter, uncheck the "Scalar values" check box to
allow this parameter to take type-dependent values.
#+end_quote

[[./figures/tutorial_ReSetup.png]]

The next parameter is the "become uninfectious rate" parameter, representing
the rate at which infected individuals become uninfectious. As this parameter
is predominantly a function of the pathogen rather than location, we will
assume it takes a single value which  is shared among types. Furthermore,
as influenza infections often take a relatively short time to pass, we will initialise
this value to correspond to an infectious period of 1 week.

#+begin_quote
Find the Become Uninfectious Rate parameter, double-click the starting value and
change it to 52 (per year).  *Be sure to press "Enter" to cause BEAUti to accept
the new value.*
#+end_quote

[[./figures/tutorial_BUSetup.png]]

Now let's consider the Sampling Proportion parameter. This parameter requires
special care, as the sampling assumptions of the model can dramatically influence
inference results.  In our case we want (a) to allow for type-dependent sampling proportions
and (b) to fix the sampling proportion to zero for all times earlier than the first
sample.

In BDMM-Prime, this configuration is reasonably simple to configure.
#+begin_quote
Find the Sampling Proportion parameter, then follow these steps:
1. Uncheck the "Scalar value" checkbox to allow for type-dependent values.
2. Check "Display visualization" to see the distribution of sample times relative
   to the most recent sample.
3. Set the "Number of change times" value to 1.  (Notice the appearance of the epoch
   boundary marker on the visualisation.)
4. Set the change time for the boundary between epoch 1 and 2 to 5.7. (Notice that this
   value ensures the oldest sample remains in epoch 1.)
5. In the "Values" table, double-click the entries corresponding to
   the sampling proportion values for Epoch 2 and change each to 0.0,
   remembering to press "enter" after each change.
#+end_quote

[[./figures/tutorial_SampSetup.png]]

Finally, we need to define how lineages change type in our multi-type
model.  Like BDMM, BDMM-Prime allows both direct "migration" as well
as "birth to a new type" styles of multi-type models.  Given we are
modelling the movement of infected influenza hosts, migration suits
our situtation best.  We also acknowledge that migration rates between
different pairs of types/locations may be different from one another.

To incorporate these decisions, follow these instructions:
#+begin_quote
Find the Migration Rate parameter, then:
1. Double-click on the value and change it from it's default to the smaller
   rate of 0.1 (any infected individual has a 10% chance of moving between
   the two locations in any given year).
2. Uncheck the "Scalar values" checkbox to allow rates to vary between pairs of
   locations.
3. Check the "Estimate values" checkbox to ensure these migration
   rates are estimated.
#+end_quote

[[./figures/tutorial_migSetup.png]]

*** Setting the remaining parameter priors
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-Setting-the-remaining-parameter-priors
:END:

With the multi-type model is configured, what remains is to define priors for
each of the parameters we're estimating.  As in any ther BEAST analysis, this
includes parameters each of the various site, clock and phylodynamic models we've
configured.

Rather than describe in detail how to set these priors, here we will simply list
a set of simple priors which can be used for the purpose of demonstrating the
model. (Detailed information on setting up priors in BEAUti can be found elsewhere, see
for instance the [[https://taming-the-beast.org/tutorials/Prior-selection/][Prior Selection tutorial]] on the Taming the BEAST website.

#+begin_quote
Set the following parameter priors:
|------------------------------------------+------------------|
| Parameter                                | Prior            |
|------------------------------------------+------------------|
| =ReEpi.t:h3n2_2deme=                     | LogNormal(0,0.5) |
| =becomeUninfectiousRateEpi.t:h3n2_2deme= | 1/X              |
| =migrationRateEpi.t:h3n2_2deme=          | Exp(0.1)         |
| =samplingProportionEpi.t:h3n2_2deme=     | Unif(0,1)        |
|------------------------------------------+------------------|
#+end_quote

*** MCMC and Logging setup
:PROPERTIES:
:CUSTOM_ID: id-3-Setting-up-your-first-analysis-Getting-Started-MCMC-and-Logging-setup
:END:

Finall, we need to configure the details of the MCMC chain to run and the trees and
parameters to be logged.

#+begin_quote
Switch to the MCMC panel and change the Chain Length to 1000000 ($10^6$).
#+end_quote

[[./figures/tutorial_MCMC.png]]

For this demonstration we will leave everything else as-is. Notice however that
beyond the regular trace, screen and tree logs, there are three additional loggers.
We'll discuss these in more detail 

- typedTreeLogger :: A logger which uses stochastic mapping to
  generate samples from the marginal posterior over edge-typed trees,
  which amount to hypotheses about when and where each type change
  occurred on the tree.
  
- nodeTypedTreeLogger :: Similar to typedTreeLogger, but the sampled trees
  includes type information only at internal nodes.  Used to build summary trees
  with BEAST 2's TreeAnnotator utility.

- trajLogger :: A logger which uses particle filtering to log samples
  from the marginal posterior of multi-type population trajectories.
  
Each of these loggers is optional, and can be enabled/disabled by
clicking the arrow buttons to their left and checking/unchecking the
"Enable logger" box.  By default edge-typed tree logging is enabled,
but trajectory logging is disabled.

** Saving and Running the analysis
:PROPERTIES:
:CUSTOM_ID: id-2-Getting-Started-Saving-and-Running-the-analysis
:END:

Now we can save and run the analysis:
#+begin_quote
1. Open the "File" menu and select "Save". Choose a sensible location
   in your filesystem in which to save the analysis file, which should
   be named ~h3n2.xml~.  (By default the location will be the
   examples/ folder from which you loaded the sequences: you'll want
   to choose somewhere else, perhaps a folder on your desktop.)
2. Launch BEAST, select the ~h3n2.xml~ as the input file, then run the analysis.
#+end_quote

** Processing the results
:PROPERTIES:
:CUSTOM_ID: id-2-Getting-Started-Processing-the-results
:END:

Once the analysis is complete, we can have a look at the results.  (If
you're impatient, note that BEAST 2 regularly appends to output files
and that these files can also be studied in the followin way while
BEAST is still running.)  The output files which should have been produced are

- =h3n2.log= :: The parameter log file,
- =h3n2-h3n2_2deme.trees= :: A tree log file with no types marked,
- =h3n2.h3n2_2deme.typed.trees= :: A tree log file with all ancestral types and type changes marked,
- =h3n2.h3n2_2deme.typed.node.trees= :: A tree log  file with only types at ancestral nodes marked.

Firstly, let's open the parameter log file using the Tracer app (distributed with BEAST 2).

#+begin_quote
Open Tracer, select "Import trace file..." from the File menu.  Use the interface
to explor the marginal distributions as estimated from the MCMC samples.
#+end_quote

We see immediately that the ESS for many of the parameters is low,
which is unsurprising for such a short run.  However, the focus here is on
understanding how BDMM-Prime analysis results are presented in the log file rather
than trying to seriously interpret the results.

With this in mind, see firstly that each of the estimated parameters including ~Re~,
~becomeUninfectiousRate~ and ~migrationRate~ are represented in the log file
by "ReSPEpi", "becomeUninfectiousRateSPEpi" and "migrationRateSPEpi".  The "SP"
indicates that the parameter is a "Skyline parameter" (as opposed to BEAST 2's
regular RealParameter).

Also notice that when we have allowed type-specific (not scalar) values, as
in the case of ~Re~, the parameter name appears multiple times with a suffix
indicating the type associated with the specific value.

Finally, notice that when we have allowed the parameter to change at a particular
time, each epoch/interval receives its own set of parameter values. In this case
an additional ".endtime" value is also recorded, which identifies the end time
of a specific interval, measured relative to the start time of the process.

[[./analyses/tracer.png]]

The untyped tree log file can be dealt with in the same way as all beast tree files.
However we also have tree log files containing sampled ancestral type information
which we can use to learn about the type change process as it occurred on the tree.
In this influenza case, this contains potentially interesting information about
global influenza spatial dynamics.

While in principle the .typed.trees file contains the most detailed information,
it is easier to jointly summarize these ancestral dynamics with the tree itself
by applying the TreeAnnotator program to the .typed.node.trees file.

#+begin_quote
1. Open TreeAnnotator (distributed with BEAST 2)
2. For the node heights option, select "mean heights". (This produces
   less biased estimates of divergence times than the default.)
3. Select =h3n2.h3n2_2deme.typed.node.trees= as the input file.
4. Choose an appropriate output file name (e.g. =summary.tree=)
5. Click "Run".
#+end_quote

[[./analyses/treeannotator.png]]

Once complete, the
#+begin_quote
1. Navigate to https://icytree.org in a javascript-enabled web browser.
2. Select "Load from file" from the File menu and select the =summary.tree= file
   produced by TreeAnnotator.
3. From the Style menu choose "colour nodes by" and then "type".
3. Also from the Style menu choose "Node height error bars" and then "=height_95%_HPD=".
#+end_quote

This should produce something quite like the figure below:

[[./analyses/summary_tree.png]]

* Model specification using BEAUti
:PROPERTIES:
:CUSTOM_ID: id-1-Model-specification-using-BEAUti
:END:

In the following section we'll delve into the BDMM-Prime interface in
more detail.  Unlike the tutorial, this section is written in a more
general way rather than considering a specific data set. However, we
will occasionally refer to the example in the tutorial, so it would be
helpful if you'd worked through that first before proceeding to the
sections below.

** The BDMM-Prime Tree Prior
:PROPERTIES:
:CUSTOM_ID: id-2-Model-specification-using-BEAUti-The-BDMM-Prime-Tree-Prior
:END:

Using BDMM-Prime amounts to selecting the BDMM-Prime tree prior when
choosing priors for a particular analysis. This is a very flexible
tree prior, encompassing a wide variety of single- and multi-type
birth-death-sampling models, allowing for fixed-rate and
time-dependent-rate analyses.

For earlier packages such as BDSKY and BDMM, the inputs to the relevant tree priors
were organised in very flat manner, with the tree prior directly taking

In contrast, inputs to the BDMM-Prime tree prior are organised in a hierarchical way


** Parameterizations
:PROPERTIES:
:CUSTOM_ID: id-2-Model-specification-using-BEAUti-Parameterizations
:END:

Unlike packages such as BDMM and BDSKY which provide a fixed set of
(often overlapping) parameterizations to work with (even at the XML level),
BDMM-Prime provides a plug-in mechanism for 

*** Canonical
:PROPERTIES:
:CUSTOM_ID: id-3-Parameterizations-Model-specification-using-BEAUti-Canonical
:END:

*** Epidemiological
:PROPERTIES:
:CUSTOM_ID: id-3-Parameterizations-Model-specification-using-BEAUti-Epidemiological
:END:

*** Fossilized birth-death process
:PROPERTIES:
:CUSTOM_ID: id-3-Parameterizations-Model-specification-using-BEAUti-Fossilized-birth-death-process
:END:

** Skyline Parameters
:PROPERTIES:
:CUSTOM_ID: id-2-Model-specification-using-BEAUti-Skyline-Parameters
:END:

*** Vector Parameters
:PROPERTIES:
:CUSTOM_ID: id-3-Skyline-Parameters-Model-specification-using-BEAUti-Vector-Parameters
:END:

*** Matrix Parameters
:PROPERTIES:
:CUSTOM_ID: id-3-Skyline-Parameters-Model-specification-using-BEAUti-Matrix-Parameters
:END:

*** Sample distribution visualisation
:PROPERTIES:
:CUSTOM_ID: id-3-Skyline-Parameters-Model-specification-using-BEAUti-Sample-distribution-visualisation
:END:

** TypeTraitSet and StartTypeProbs
:PROPERTIES:
:CUSTOM_ID: id-2-Model-specification-using-BEAUti-TypeTraitSet-and-StartTypeProbs
:END:

** Miscellaneous inputs
:PROPERTIES:
:CUSTOM_ID: id-2-Model-specification-using-BEAUti-Miscellaneous-inputs
:END:

* Sampling Latent Variables
:PROPERTIES:
:CUSTOM_ID: id-1-Sampling-Latent-Variables
:END:

** Tree edge types and type transitions
:PROPERTIES:
:CUSTOM_ID: id-2-Sampling-Latent-Variables-Tree-edge-types-and-type-transitions
:END:

** Multi-type trajectories
:PROPERTIES:
:CUSTOM_ID: id-2-Sampling-Latent-Variables-Multi-type-trajectories
:END:

* XML reference
:PROPERTIES:
:CUSTOM_ID: id-1-XML-reference
:END:

** Model definition
:PROPERTIES:
:CUSTOM_ID: id-2-XML-reference-Model-definition
:END:

** Special operators
:PROPERTIES:
:CUSTOM_ID: id-2-XML-reference-Special-operators
:END:

* License
:PROPERTIES:
:CUSTOM_ID: id-1-License
:END:

BDMM-Prime is free software.  It is made available under the terms of
the [[https://www.gnu.org/licenses/gpl-3.0.html][GNU General Public Licence version 3]]. Unlike many software
licenses, this is designed to give you, the user, the freedom to use
the software as you wish - including making and publishing
modifications - provided you also extend the same freedoms to other users.


# Evaluate the following block to update section IDs
#
#+begin_src emacs-lisp :cache no :exports none :results none
  (org-map-entries
     (lambda ()
       (pcase (org-heading-components)
         (`(,level ,_ ,_ ,_ ,title ,_)
          (org-set-property "CUSTOM_ID"
                            (string-replace " " "-"
                             (string-join `("id"
                                            ,(number-to-string level)
                                            ,@(reverse (org-get-outline-path))
                                            ,title)
                                          "-")))))))
#+end_src

# Local Variables:
# org-html-validation-link: nil
# org-html-htmlize-output-type: css
# End:
